tools/lkl/lib/rump-host.c

```diff
@@ -0,0 +1,672 @@
+{
+	return 0;
+}
+
+int rump___sysimpl_socket30(int i, int j, int k)
+{
+	return 0;
+}
+
+int rump___sysimpl_unmount(const char *str, int i)
+{
+	return 0;
+}
+
+void __assert13(const char *file, int line, const char *function,
+		const char *failedexpr)
+{
+}
+
+int rump___sysimpl_close(int fd)
+{
+	return -1;
+}
+
+int rump___sysimpl_ioctl(int fd, u_long com, void *data)
+{
+	return -1;
+}
+
+int rump___sysimpl_mkdir(const char *path, mode_t mode)
+{
+	return -1;
+}
+
+int rump___sysimpl_open(const char *name, int flags, ...)
+{
+	return -1;
+}
+
+#endif /* RUMP_TEMP_STUB */
+
+#ifndef RUMPRUN
+static int fd_get_capacity(struct lkl_disk disk, unsigned long long *res)
+{
+	off_t off;
+
+	off = lseek(disk.fd, 0, SEEK_END);
+	if (off < 0)
+		return -1;
+
+	*res = off;
+	return 0;
+}
+
+static int blk_request(struct lkl_disk disk, struct lkl_blk_req *req)
+{
+	int err = 0;
+	struct iovec *iovec = (struct iovec *)req->buf;
+
+	/* TODO: handle short reads/writes */
+	switch (req->type) {
+	case LKL_DEV_BLK_TYPE_READ:
+		err = preadv(disk.fd, iovec, req->count, req->sector * 512);
+		break;
+	case LKL_DEV_BLK_TYPE_WRITE:
+		err = pwritev(disk.fd, iovec, req->count, req->sector * 512);
+		break;
+	case LKL_DEV_BLK_TYPE_FLUSH:
+	case LKL_DEV_BLK_TYPE_FLUSH_OUT:
+		err = fsync(disk.fd);
+		break;
+	default:
+		return LKL_DEV_BLK_STATUS_UNSUP;
+	}
+
+	if (err < 0)
+		return LKL_DEV_BLK_STATUS_IOERR;
+
+	return LKL_DEV_BLK_STATUS_OK;
+}
+
+struct lkl_dev_blk_ops lkl_dev_blk_ops = {
+	.get_capacity = fd_get_capacity,
+	.request = blk_request,
+};
+
+struct lkl_netdev_rumpfd {
+	struct lkl_netdev dev;
+	/* TAP device */
+	int fd;
+};
+
+static int rump_net_tx(struct lkl_netdev *nd,
+		       struct iovec *iov, int cnt)
+{
+	struct lkl_netdev_rumpfd *nd_rumpfd =
+		container_of(nd, struct lkl_netdev_rumpfd, dev);
+	int ret;
+
+	do {
+		ret = writev(nd_rumpfd->fd, iov, cnt);
+	} while (ret == -1 && (errno == EINTR));
+
+	if (ret < 0)
+		lkl_perror("write to rump fd netdev fails", errno);
+
+	return ret;
+}
+
+static int rump_net_rx(struct lkl_netdev *nd,
+		       struct iovec *iov, int cnt)
+{
+	struct lkl_netdev_rumpfd *nd_rumpfd =
+		container_of(nd, struct lkl_netdev_rumpfd, dev);
+	int ret;
+
+	do {
+		ret = readv(nd_rumpfd->fd, iov, cnt);
+	} while (ret == -1 && errno == EINTR);
+
+	if (ret <= 0)
+		return -1;
+
+	return ret;
+}
+
+static int rump_net_poll(struct lkl_netdev *nd)
+{
+	struct lkl_netdev_rumpfd *nd_rumpfd =
+		container_of(nd, struct lkl_netdev_rumpfd, dev);
+	struct pollfd pfd = {
+		.fd = nd_rumpfd->fd,
+		.events = POLLIN | POLLPRI | POLLOUT
+	};
+	int ret = 0;
+
+
+	while (1) {
+		int err = poll(&pfd, 1, -1);
+
+		if (err < 0 && errno == EINTR)
+			continue;
+		if (err > 0)
+			break;
+	}
+
+	if (pfd.revents & (POLLHUP | POLLNVAL))
+		return -1;
+
+	if (pfd.revents & POLLIN)
+		ret |= LKL_DEV_NET_POLL_RX;
+	if (pfd.revents & POLLOUT)
+		ret |= LKL_DEV_NET_POLL_TX;
+
+	return ret;
+}
+
+struct lkl_dev_net_ops rumpfd_ops = {
```

Comment:
The LKL way of doing things is to export the raw hardware I/O to the underlying/hosting system.  In the BMK rumpkernel implementation, you actually use the devices within the RK for I/O.  Given that you've implemented the PCI rumpuser interface, are LKL device drivers supported?

Reply:
this particular piece of code (`rumpfd_ops`) is used by https://github.com/libos-nuse/frankenlibc, which is a variant of rumprun unikernel.  and you're right: with rumprun bmk LKL can use drivers/net/virtio-net.c or drivers/block/virtio-blk.c.  I have tried e1000 driver of Linux with qemu e1000 emulation over LKL.  Though recently rumprun LKL is not well maintained (I hope I do as early as possible).

Reply:
I'm familiar with rumprun (we have our own hacked up version with a backend for our own OS).  I was not sure that you'd be able to run native drivers (e.g. e1000) since I didn't see the virtual to physical rumpuser function (which is used in `bmk`, for example to enable devices to understand how to interact with DMA).

Great to hear that devices are at least possible, if not supported!  

I was unaware of the mainlining initiative so I certainly understand why this patchset is currently unmaintained.